azure-batch/azure/batch/custom/patch.py

```diff
@@ -0,0 +1,287 @@
+                    # exception and have it work in all cases where tasks are well behaved
+                    # Behavior retries as a smaller chunk and
+                    # appends extra tasks to queue to be picked up by another thread .
+                    self._tasks_to_add.extendleft(chunk_tasks_to_add[midpoint:])
+                    self._bulk_add_tasks(results_queue, chunk_tasks_to_add[:midpoint])
+            elif 500 <= e.response.status_code <= 599:
+                self._tasks_to_add.extendleft(chunk_tasks_to_add)
+            else:
+                results_queue.appendleft(e)
+        except Exception as e:  # pylint: disable=broad-except
+            results_queue.appendleft(e)
+        else:
+            try:
+                add_collection_response = add_collection_response.output
+            except AttributeError:
+                pass
+
+            for task_result in add_collection_response.value:  # pylint: disable=no-member
+                if task_result.status == TaskAddStatus.server_error:
+                    # Server error will be retried
+                    with self._pending_queue_lock:
+                        for task in chunk_tasks_to_add:
+                            if task.id == task_result.task_id:
+                                self._tasks_to_add.appendleft(task)
+                elif (task_result.status == TaskAddStatus.client_error
+                        and not task_result.error.code == "TaskExists"):
+                    # Client error will be recorded unless Task already exists
+                    self._failures.appendleft(task_result)
+                else:
+                    results_queue.appendleft(task_result)
+
+    def task_collection_thread_handler(self, results_queue):
+        """Main method for worker to run
+
+        Pops a chunk of tasks off the collection of pending tasks to be added and submits them to be added.
+
+        :param collections.deque results_queue: Queue for worker to output results to
+        """
+        while self._tasks_to_add and not self._has_early_termination_error:
+            max_tasks = self._max_tasks_per_request  # local copy
+            chunk_tasks_to_add = []
+            with self._pending_queue_lock:
+                while len(chunk_tasks_to_add) < max_tasks and self._tasks_to_add:
+                    chunk_tasks_to_add.append(self._tasks_to_add.pop())
+
+            if chunk_tasks_to_add:
+                self._bulk_add_tasks(results_queue, chunk_tasks_to_add)
+
+
+        # Only define error if all threads have finished and there were failures
+        with self._error_lock:
+            if threading.active_count() == 1 and self._failures:
+                self.error = errors.CreateTasksErrorException(
+                    "One or more tasks failed to be added",
+                    self._failures,
+                    self._tasks_to_add)
```

Comment:
It seems to me that self._tasks_to_add will only ever have something in it if **_both_** a client error occurred **_and_** an early termination due to an over-sized request body - is that the case?

Reply:
At this line, self._tasks_to_add can be non-empty iff there was an early termination. 
                   self._failures can be non-empty iff there were client-errors to report back to the user.

self._tasks_to_add => contains tasks not yet added

Reply:
Yes - so as I said. I'm not saying it's an issue - just clarifying.

Reply:
You are correct.  Can probably remove from the error contents then because the reason we early terminate is because we want to communicate the original error to the user.